File: frontend/src/api/chapter.ts
import request from './request'

// 章节数据类型
export interface Chapter {
  id: number
  project_id: number
  title: string
  content: string
  summary: string | null
  word_count: number
  sort_order: number
  status: string
  pov_character: string | null
  created_at: string
  updated_at: string
}

export interface CreateChapterData {
  title: string
  content?: string
  sort_order?: number
  summary?: string
  status?: string
  pov_character?: string
}

export interface UpdateChapterData {
  title?: string
  content?: string
  sort_order?: number
  summary?: string
  status?: string
  pov_character?: string
}

export interface ChapterOrder {
  id: number
  sort_order: number
}

// 获取项目的所有章节
export async function getChapters(projectId: number): Promise<Chapter[]> {
  return request.get<Chapter[]>(`/projects/${projectId}/chapters/`)
}

// 创建章节
export async function createChapter(projectId: number, data: CreateChapterData): Promise<Chapter> {
  return request.post<Chapter>(`/projects/${projectId}/chapters/`, data)
}

// 获取单个章节
export async function getChapter(projectId: number, chapterId: number): Promise<Chapter> {
  return request.get<Chapter>(`/projects/${projectId}/chapters/${chapterId}/`)
}

// 更新章节
export async function updateChapter(
  projectId: number,
  chapterId: number,
  data: UpdateChapterData
): Promise<Chapter> {
  return request.put<Chapter>(`/projects/${projectId}/chapters/${chapterId}/`, data)
}

// 删除章节
export async function deleteChapter(projectId: number, chapterId: number): Promise<void> {
  return request.delete(`/projects/${projectId}/chapters/${chapterId}/`)
}

// 批量删除章节
export async function batchDeleteChapters(projectId: number, chapterIds: number[]): Promise<void> {
  return request.post(`/projects/${projectId}/chapters/batch-delete/`, { ids: chapterIds })
}

// 重新排序章节
export async function reorderChapters(projectId: number, orders: ChapterOrder[]): Promise<void> {
  return request.post(`/projects/${projectId}/chapters/reorder/`, { orders })
}

// ========== 章节版本历史相关 ==========

export interface ChapterVersion {
  id: number
  chapter_id: number
  version_number: number
  title: string
  word_count: number
  change_summary: string | null
  created_at: string
}

export interface ChapterVersionDetail extends ChapterVersion {
  content: string
}

export interface RestoreResult {
  message: string
  chapter: Chapter
}

// 手动保存章节版本
export async function saveChapterVersion(
  projectId: number,
  chapterId: number
): Promise<ChapterVersion> {
  return request.post<ChapterVersion>(
    `/projects/${projectId}/chapters/${chapterId}/versions/save`
  )
}

// 获取章节版本列表
export async function getChapterVersions(
  projectId: number,
  chapterId: number
): Promise<ChapterVersion[]> {
  return request.get<ChapterVersion[]>(`/projects/${projectId}/chapters/${chapterId}/versions/`)
}

// 获取版本详情
export async function getChapterVersion(
  projectId: number,
  chapterId: number,
  versionId: number
): Promise<ChapterVersionDetail> {
  return request.get<ChapterVersionDetail>(
    `/projects/${projectId}/chapters/${chapterId}/versions/${versionId}/`
  )
}

// 恢复到指定版本
export async function restoreChapterVersion(
  projectId: number,
  chapterId: number,
  versionId: number
): Promise<RestoreResult> {
  return request.post<RestoreResult>(
    `/projects/${projectId}/chapters/${chapterId}/versions/${versionId}/restore/`
  )
}

File: backend/app/schemas/chapter.py
"""
章节 Pydantic 模式
"""
from datetime import datetime
from typing import Optional, List

from pydantic import BaseModel, ConfigDict


class ChapterCreate(BaseModel):
    """创建章节时的请求体"""
    title: str
    content: str = ""
    summary: Optional[str] = None
    sort_order: Optional[int] = None  # 不传则自动追加到末尾
    status: str = "draft"
    pov_character: Optional[str] = None


class ChapterUpdate(BaseModel):
    """更新章节时的请求体（所有字段均为 Optional）"""
    title: Optional[str] = None
    content: Optional[str] = None
    summary: Optional[str] = None
    sort_order: Optional[int] = None
    status: Optional[str] = None
    pov_character: Optional[str] = None


class ChapterResponse(BaseModel):
    """章节详情响应"""
    model_config = ConfigDict(from_attributes=True)

    id: int
    project_id: int
    title: str
    content: str
    summary: Optional[str] = None
    sort_order: int
    word_count: int
    status: str
    pov_character: Optional[str] = None
    created_at: datetime
    updated_at: Optional[datetime] = None


class ChapterBatchDeleteRequest(BaseModel):
    """批量删除请求体"""
    ids: list[int]


class ChapterReorderItem(BaseModel):
    """批量重排序时单条记录"""
    id: int
    sort_order: int


class ChapterReorderRequest(BaseModel):
    """批量重排序请求体"""
    orders: list[ChapterReorderItem]


# ========== 章节版本历史相关 ==========

class ChapterVersionResponse(BaseModel):
    """章节版本响应"""
    model_config = ConfigDict(from_attributes=True)

    id: int
    chapter_id: int
    version_number: int
    title: str
    word_count: int
    change_summary: Optional[str] = None
    created_at: datetime


class ChapterVersionDetail(BaseModel):
    """章节版本详情（包含内容）"""
    model_config = ConfigDict(from_attributes=True)

    id: int
    chapter_id: int
    version_number: int
    title: str
    content: str
    word_count: int
    change_summary: Optional[str] = None
    created_at: datetime


class ChapterVersionRestore(BaseModel):
    """版本恢复响应"""
    message: str
    chapter: "ChapterResponse"
